src/lib.rs

```diff
@@ -95,130 +95,165 @@ fn impl_with_generator<G: BodyGenerator>(ast: &syn::DeriveInput, gen: G) -> quot
-        owned_lifetime_params.chain(owned_type_params).collect::<Vec<_>>()
+    fn quote_rhs_params(&self, ast: &syn::DeriveInput) -> Vec<proc_macro2::TokenStream> {
+        let owned_lifetime_params = ast.generics.lifetimes().map(|_| quote! { 'static });
+        let owned_type_params = ast.generics.type_params().map(|ty| {
+            let ident = &ty.ident;
+            quote! { #ident }
+        });
+        owned_lifetime_params
+            .chain(owned_type_params)
+            .collect::<Vec<_>>()
     }
 
-    fn visit_struct(&self, data: &syn::VariantData) -> quote::Tokens;
-    fn visit_enum_data(&self, variant: quote::Tokens, data: &syn::VariantData) -> quote::Tokens;
-    fn combine_impl(&self, borrows: quote::Tokens, name: &syn::Ident, rhs_params: quote::Tokens, owned: quote::Tokens, body: quote::Tokens) -> quote::Tokens;
+    fn visit_struct(&self, data: &syn::DataStruct) -> proc_macro2::TokenStream;
+    fn visit_enum_data(
+        &self,
+        variant: proc_macro2::TokenStream,
+        data: &syn::Variant,
+    ) -> proc_macro2::TokenStream;
+    fn combine_impl(
+        &self,
+        borrows: proc_macro2::TokenStream,
+        name: &syn::Ident,
+        rhs_params: proc_macro2::TokenStream,
+        owned: proc_macro2::TokenStream,
+        body: proc_macro2::TokenStream,
+    ) -> proc_macro2::TokenStream;
 }
 
 struct IntoOwnedGen;
 
 impl BodyGenerator for IntoOwnedGen {
-
-    fn visit_struct(&self, data: &syn::VariantData) -> quote::Tokens {
-        match *data {
-            syn::VariantData::Struct(ref body) => {
-                let fields = body.iter()
-                    .map(|field| {
-                        let ident = field.ident.as_ref().unwrap();
-                        let field_ref = quote! { self.#ident };
-                        let code = FieldKind::resolve(field).move_or_clone_field(&field_ref);
-                        quote! { #ident: #code }
-                    })
-                    .collect::<Vec<_>>();
-                quote! { { #(#fields),* } }
-            },
-            syn::VariantData::Tuple(ref body) => {
-                let fields = body.iter()
-                    .enumerate()
-                    .map(|(index, field)| {
-                        let index = syn::Ident::from(index);
-                        let index = quote! { self.#index };
-                        FieldKind::resolve(field).move_or_clone_field(&index)
-                    })
-                    .collect::<Vec<_>>();
-                quote! { ( #(#fields),* ) }
-            },
-            syn::VariantData::Unit => {
-                quote! {}
-            }
+    fn visit_struct(&self, data: &syn::DataStruct) -> proc_macro2::TokenStream {
+        let fields_are_named = data
+            .fields
+            .iter()
+            .next()
+            .map(|field| field.ident.is_some())
+            .expect("empty struct? yes, just default to true or false here");
```

Comment:
I don't think I understand the meaning of this expect (comparing only to "my" ancient code here). It'd appear that the previous version returned `quote! {}` for unit structs.

Maybe a test case was missing for this since I had not discovered (or it didn't yet exist) [`trybuild`](https://github.com/dtolnay/trybuild) ... Or maybe it was just missing?

I'll try this before merging, maybe PR your PR.

Sounds like a 

```suggestion
            .unwrap_or(false); // handle empty unit structs as tuples
```

Would work here, but need to test it locally.

Reply:
Yes, managed to read this correctly:

```
diff --git a/tests/tuple_struct.rs b/tests/tuple_struct.rs
index 05e7467..f0b9d36 100644
--- a/tests/tuple_struct.rs
+++ b/tests/tuple_struct.rs
@@ -21,6 +21,9 @@ struct Car<'a>(std::borrow::Cow<'a, str>);
 #[derive(IntoOwned)]
 struct Dar<'a>(borrow::Cow<'a, str>);

+#[derive(IntoOwned)]
+struct Far;
+
 #[test]
 fn tuple_struct() {
     let non_static_string: String = "foobar".to_string();
```

Will hit the `expect`.

The previous work around for this works, I'll just send you the patch.

Reply:
Note I think the suggested unwrap_ot(false) does not work with the test case I thought of, which was missing before the upgrade. I wonder if you got any notification on https://github.com/hoodie/derive-into-owned/pull/1 ?